base/linalg/uniformscaling.jl

```diff
@@ -24,14 +24,63 @@ one{T}(J::UniformScaling{T}) = one(UniformScaling{T})
 zero{T}(::Type{UniformScaling{T}}) = UniformScaling(zero(T))
 zero{T}(J::UniformScaling{T}) = zero(UniformScaling{T})
 
-+(J1::UniformScaling, J2::UniformScaling) = UniformScaling(J1.λ+J2.λ)
-+{T}(B::BitArray{2},J::UniformScaling{T}) = bitunpack(B) + J
-+(J::UniformScaling, B::BitArray{2})      = J + bitunpack(B)
-+(J::UniformScaling, A::AbstractMatrix)   = A + J
-+(J::UniformScaling, x::Number)           = J.λ + x
-+(x::Number, J::UniformScaling)           = x + J.λ
-
-function +{TA,TJ}(A::AbstractMatrix{TA}, J::UniformScaling{TJ})
+(+)(J1::UniformScaling, J2::UniformScaling) = UniformScaling(J1.λ+J2.λ)
+(+){T}(B::BitArray{2},J::UniformScaling{T}) = bitunpack(B) + J
+(+)(J::UniformScaling, B::BitArray{2})      = J + bitunpack(B)
+(+)(J::UniformScaling, A::AbstractMatrix)   = A + J
+(+)(J::UniformScaling, x::Number)           = J.λ + x
+(+)(x::Number, J::UniformScaling)           = x + J.λ
+
+(-)(J::UniformScaling)                      = UniformScaling(-J.λ)
+(-)(J1::UniformScaling, J2::UniformScaling) = UniformScaling(J1.λ-J2.λ)
+(-)(B::BitArray{2}, J::UniformScaling)      = bitunpack(B) - J
+(-)(J::UniformScaling, B::BitArray{2})      = J - bitunpack(B)
+(-)(J::UniformScaling, x::Number)           = J.λ - x
+(-)(x::Number, J::UniformScaling)           = x - J.λ
+
+for (t1, t2) in ((:UnitUpperTriangular, :UpperTriangular),
+                 (:UnitLowerTriangular, :LowerTriangular))
+    for op in (:+,:-)
+        @eval begin
+            ($op)(UL::$t2, J::UniformScaling) = ($t2)(($op)(UL.data, J))
+
+            function ($op)(UL::$t1, J::UniformScaling)
+                ULnew = copy_oftype(UL.data, promote_type(eltype(UL), eltype(J)))
+                for i = 1:size(ULnew, 1)
+                    ULnew[i,i] = ($op)(1, J.λ)
+                end
+                return ($t2)(ULnew)
+            end
+        end
+    end
+
+    function (-)(J::UniformScaling, UL::Union{UpperTriangular,UnitUpperTriangular})
+        ULnew = similar(UL, promote_type(eltype(J), eltype(UL)))
+        n = size(ULnew, 1)
+        ULold = UL.data
+        for j = 1:n
+            for i = 1:j - 1
+                ULnew[i,j] = -ULold[i,j]
+            end
+            ULnew[j,j] = J.λ - ifelse(isa(UL, UnitUpperTriangular), 1, ULold[j,j])
+        end
+        return UpperTriangular(ULnew)
+    end
+    function (-)(J::UniformScaling, UL::Union{LowerTriangular,UnitLowerTriangular})
+        ULnew = similar(UL, promote_type(eltype(J), eltype(UL)))
+        n = size(ULnew, 1)
+        ULold = UL.data
+        for j = 1:n
+            ULnew[j,j] = J.λ - ifelse(isa(UL, UnitLowerTriangular), 1, ULold[j,j])
+            for i = j + 1:n
+                ULnew[i,j] = -ULold[i,j]
+            end
+        end
+        return LowerTriangular(ULNew)
```

Comment:
I believe it does. It is defined in line 70, but it should probably be `similar(UL.data`. I'll look into that. The first problem is to figure how to modify the signature such that Julia considers a `XTriangularMatrix` more specific than an `AbstractMatrix`.